src/lib/rpc/bestRpcTracker.ts

```diff
@@ -0,0 +1,346 @@
+  responseTime: number;
+  blockNumber: number | null;
+  timestamp: Date;
+};
+
+type ProviderData = {
+  url: string;
+  provider: Provider;
+};
+
+type RpcTrackerState = {
+  [chainId: number]: {
+    lastUsage: Date | null;
+    currentBestProviderUrl: string;
+    providers: {
+      [providerUrl: string]: ProviderData;
+    };
+  };
+};
+
+let trackingIntervalId: number | null = null;
+let trackerState: RpcTrackerState | null = null;
+
+function initRpcTracking() {
+  trackerState = initTrackerState();
+
+  if (trackingIntervalId) {
+    clearInterval(trackingIntervalId);
+  }
+  measureRpcData({ warmUp: true });
+  trackingIntervalId = window.setInterval(() => measureRpcData(), PROBE_INTERVAL);
+}
+
+function measureRpcData({ warmUp = false } = {}) {
+  if (!trackerState) {
+    throw new Error("RPC tracker state is not initialized");
+  }
+
+  entries(trackerState).forEach(async ([chainIdRaw, chainTracker]) => {
+    const chainId = Number(chainIdRaw);
+    const providers = Object.values(chainTracker.providers);
+
+    const isUnusedChain =
+      !chainTracker.lastUsage ||
+      differenceInMilliseconds(Date.now(), chainTracker.lastUsage) > DISABLE_UNUSED_TRACKING_TIMEOUT;
+    const isTrackingEnabled = (warmUp || !isUnusedChain) && providers.length > 1;
+
+    if (!isTrackingEnabled) {
+      return;
+    }
+
+    const probePromises = providers.map((providerInfo) => {
+      return probeRpc(chainId, providerInfo.provider, providerInfo.url);
+    });
+
+    const probeResults = await Promise.all(probePromises);
+    const successProbeResults = probeResults.filter((probe) => probe.isSuccess);
+
+    if (!successProbeResults.length) {
+      setCurrentProvider(chainId, getFallbackRpcUrl(chainId));
+
+      return;
+    }
+
+    const probeResultsByBlockNumber = orderBy(successProbeResults, ["blockNumber"], ["desc"]);
+    let bestBlockNumberProbe = probeResultsByBlockNumber[0];
+
+    const probeStats = probeResultsByBlockNumber.map((probe, i, arr) => {
+      let isValid = probe.isSuccess;
+
+      const bestBlockNumber = bestBlockNumberProbe.blockNumber;
+      const currProbeBlockNumber = probe.blockNumber;
+      const nextProbeBlockNumber = arr[i + 1]?.blockNumber;
+
+      // Rare case when RPC returned a block number from the far future
+      if (
+        i === 0 &&
+        currProbeBlockNumber &&
+        nextProbeBlockNumber &&
+        currProbeBlockNumber - nextProbeBlockNumber > BLOCK_FROM_FUTURE_THRESHOLD
+      ) {
+        bestBlockNumberProbe = arr[i + 1];
+        isValid = false;
+      }
+
+      // If the block number is lagging behind the best one
+      if (bestBlockNumber && currProbeBlockNumber && bestBlockNumber - currProbeBlockNumber > BLOCK_LAGGING_THRESHOLD) {
+        isValid = false;
+      }
+
+      return {
+        ...probe,
+        isValid,
+      };
+    });
+
+    const bestResponseTimeValidProbe = minBy(
+      probeStats.filter((probe) => probe.isValid),
+      "responseTime"
+    );
```

Comment:
wondering if pure js would be simpler here 
```
const bestResponseTimeValidProbe = Math.min(... probeStats.filter(p => p.isValid).map(p.responseTime))
```

Reply:
The benefit of `minBy` is that it returns the object itself instead of the minimal responseTime